### ISSUE TITLE
Support redacting specific query string values in logger's default request serializer

### ISSUE THREAD

**Issue by ghost** [fastify/fastify#4980]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

Add two new fields to the default request serializer: `path` and `query`.

The `path` field would follow a similar definition to the `pathname` field from the `URL` interface [1]:

> A string containing an initial '/' followed by the path of the URL, not including the query string or fragment.

*Note: I propose `path` for Fastify simply for brevity, and because it's already commonly used interchangeably by most developers but could also use `pathname` if desired.*

The query field would simply reference the existing `req.query` object.

[1] https://developer.mozilla.org/en-US/docs/Web/API/URL/pathname

### Motivation

The current request serializer uses `req.url` [2], which includes both the path and query string parameters, which makes it impossible to redact individual query string values without either redacting the entire `url` field or providing a custom request serializer.

One example where this is extremely useful is for the OAuth authorization code flow, where both the `code` and `state` values are passed as query string parameters.

[2] https://github.com/fastify/fastify/blob/main/lib/logger.js#L52

### Example

Proposed request serializer:

```js
  req: function asReqValue (req) {
    return {
      method: req.method,
      url: req.url,
+     path: req.url.slice(0, req.url.indexOf('?')),
+     query: req.query,
      version: req.headers && req.headers['accept-version'],
      hostname: req.hostname,
      remoteAddress: req.ip,
      remotePort: req.socket ? req.socket.remotePort : undefined
    }
  },
```

To redact a specific query string parameter, you could do the following in the Fastify config object:

```js
...
  redact: ['req.url', 'req.query.secret'], // need to nuke old url field if we're using new fields
...
```

**Comment by Eomm** [comment]
I think this issue belongs to: https://github.com/pinojs/pino-std-serializers

Note that you can already implement this logic for a limited set of routes:
https://fastify.dev/docs/latest/Reference/Routes/#custom-log-serializer

> One example where this is extremely useful is for the OAuth authorization code flow, where both the code and state values are passed as query string parameters.

It is hard for me to see the security issue here: the `code` is a one-time string that is being processed while it is logged. The `state` string should be another random temp string.

**Comment by tacozilla** [comment]
Correct me if I'm wrong, but Fastify doesn't use pino's standard serializer for the request logger [1], so any changes made in pino wouldn't have any affect on this specific use case. If Fastify did use pino's default serializer, this request would actually be moot, since pino does include the `query` parameter separately.

Regarding the code, that was just one quick example, not meant to be the only purpose. Not to be pedantic, but the code is logged before it's processed, so there is an ability to exploit it if an attacker had access to live logs and was faster than the original request server, but I digress :P 

My main point was even if it's a one-time use code, it's still considered bad practice to log things like auth tokens, TOTP, etc. and it'd be nice if there was a way to use the already supported `redact` functionality.

I don't believe this would have a noticeable performance impact, and it'd be useful to have them split out in general even if we completely ignore the redact use case.  If this change is not desired, is it possible for Fastify to default to using pino's serializers for request and response, or is there a specific reason Fastify isn't using them?

[1] https://github.com/fastify/fastify/blob/main/lib/logger.js#L49-L58

**Comment by Uzlopak** [comment]
Even if you send totp or even hotp Tokens per query, it wont matter, because by nature they are only valid for a short period(totp or only once(htop).

**Comment by Eomm** [comment]
> Correct me if I'm wrong, but Fastify doesn't use pino's standard serializer for the request logger [1], so any changes made in pino wouldn't have any affect on this specific use case. If Fastify did use pino's default serializer, this request would actually be moot, since pino does include the query parameter separately.

You are right, my memory is not working anymore as before 👴🏼

Maybe we could improve add a security paragraph in the logging doc

---

Not sure if @fastify/core would like to change the default logger since it is customizable already

**Comment by jsumners** [comment]
> Not sure if @fastify/core would like to change the default logger since it is customizable already

+ https://github.com/fastify/fastify/issues/2841#issuecomment-775779962
+ https://github.com/fastify/fastify/issues/1268

**Comment by tacozilla** [comment]
@jsumners Could you elaborate a bit on how those issues are relevant to this request?

To clarify, I'm not asking for Fastify to change default parsers or change anything regarding how query strings are parsed. This request is specifically because Fastify doesn't use pino's default request logger, but rather uses its own version, which only has a single field (i.e. `req.url`) that includes both the path and query string. This makes it impossible to use the built-in redact functionality on query string parameters.

AFAICT, there are two options in the current state to achieve this:

1. Redact the entire `url` property, which is less than ideal since the path is still desirable.
2. Provide a custom request logger that is an exact copy/paste of the existing logger with 1 extra property.

Even if we ignore the redact use case for a moment, it'd still be really nice to have the query string object logged instead of the stringified version. Maybe that would have been a better way to word this request instead of specifically focusing on redacting things?

On a side note, does anyone know why Fastif doesn't use pino's default request logger?

**Comment by jsumners** [comment]
I thought I read "not sure what others feel about changing the default query string parser."

**Comment by tacozilla** [comment]
Seeing if we can close the loop on this. Three outstanding questions / clarifications:

1. **What is the best way to redact `req.url` on a single route?** It doesn't appear you can pass a redact option on a route-level basis, so you either need to redact properties globally using the `logger` property on the `Fastify` instance (definitely not ideal), or do something like the following with the `childLoggerFactory`:

```mjs
fastify.get(
  '/test',
  {
    childLoggerFactory(logger, bindings, options) {
      return logger.child(bindings, { ...options, redact: ['req.url'] });
    },
  },
  async (req, res) => 'test',
);
```

Is this accurate or am I missing an obvious way to do this in a cleaner/simpler fashion?

2. **Is the rationale documented anywhere for why Fastify uses a custom request serializer instead of pino's default request serializer?**

3. **Regardless of support for redact use case, is the Fastify team open to logging the query string in object form?** Having it split out would be nice purely from a readability/searchability standpoint for request logs and doesn't appear to have any performance issue (`req.query` is already an object on the main Fastify Request object, etc.).

Appreciate any insights anyone can share. :)

**Comment by metcoder95** [comment]
>What is the best way to redact req.url on a single route?

With the new addition, I believe using the `childLoggerFactory` its the recommended way to manage per-route logger configuration.

>Is the rationale documented anywhere for why Fastify uses a custom request serializer instead of pino's default request serializer?

I believe that it might be to log essential pieces that were not included within pinot `stdserializers`. But I lack the real answer on that, maybe someone from the team has better insights about it

>Regardless of support for redact use case, is the Fastify team open to logging the query string in object form?

I believe that this should be up to the implementation to decide which part of the request exactly to log if they want to. Having the log of the entire `url` should fulfil a generic need to see exactly what was coming at the `url` of the request, the rest must be up to the user to decide what else they want to print.

**Comment by tacozilla** [comment]
> With the new addition, I believe using the `childLoggerFactory` its the recommended way to manage per-route logger configuration.

Quick follow-up question, how is `options` populated? Regardless of what or where I set logger options elsewhere (e.g. on the main `Fastify` instance), the `options` passed to the factory function is always empty except for a single `level` key with an empty string as its value. I tried reading the docs and some of the code, but I still can't quite make sense of when or where that's getting populated.

```mjs
fastify.get(
  '/test',
  {
    childLoggerFactory(logger, bindings, options) {
      console.log(options); // this always prints => { level: '' }
      return logger.child(bindings, { ...options, redact: ['req.url'] });
    },
  },
  async () => 'test',
);
```

> I believe that it might be to log essential pieces that were not included within pinot `stdserializers`. But I lack the real answer on that, maybe someone from the team has better insights about it

I believe pino is actually a superset of Fastify, so perhaps there was a performance concern and Fastify's default request serializer was intentionally stripped down? It looks like @delvedor wrote the initial implementation, perhaps he would know, or possibly @mcollina?

**Comment by Eomm** [comment]
> What is the best way to redact req.url on a single route?

There is an example you can be inspired of: set a context with a custom serializer for `req`

https://fastify.dev/docs/latest/Reference/Routes/#custom-log-serializer

**Comment by metcoder95** [comment]
>Quick follow-up question, how is options populated? Regardless of what or where I set logger options elsewhere (e.g. on the main Fastify instance), the options passed to the factory function is always empty except for a single level key with an empty string as its value. I tried reading the docs and some of the code, but I still can't quite make sense of when or where that's getting populated.

It should provide the options object you provided through the `logger` property when creating the fastify app. If that's not happening, this most likely is a bug. Can you share a 
[Minimum Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example)?

**Comment by tacozilla** [comment]
@Eomm - Unfortunately, the custom serializer is a lot of copy/paste code to remove a single property. I think the child logger factory is a cleaner route, but are there downsides to doing that? Examples of each below:

**childLoggerFactory**
```mjs
childLoggerFactory(logger, bindings, options) {
  return logger.child(bindings, { ...options, redact: ['req.url'] });
},
```

**logSerializers**
```mjs
logSerializers: {
  req: function asReqValue (req) {
    return {
      method: req.method,
      url: req.url.slice(0, req.url.indexOf('?')),
      version: req.headers && req.headers['accept-version'],
      hostname: req.hostname,
      remoteAddress: req.ip,
      remotePort: req.socket ? req.socket.remotePort : undefined
    }
  },
}
```

@metcoder95 - Example below, let me know if you want me to create a separate issue for it, but was trying to make sure I even understood how it worked before I created an issue for something that wasn't actually a bug 😄

```mjs
import Fastify from 'fastify';

const fastify = Fastify({
  logger: {
    level: 'warn',
    redact: ['req.url'],
    msgPrefix: 'global',
  },
});

fastify.get(
  '/hello',
  {
    childLoggerFactory(logger, binding, options) {
      console.log(options); // this always logs => { level: '' }
      return logger.child(binding, options);
    },
  },
  async () => 'hello world',
);

await fastify.listen();
```

**Comment by metcoder95** [comment]
>@metcoder95 - Example below, let me know if you want me to create a separate issue for it, but was trying to make sure I even understood how it worked before I created an issue for something that wasn't actually a bug 😄

No worries at all, it is completely fine. I believe the `childLoggerFactory` should receive the logger options (either the ones passed when creating the app, the default ones or the overwrites passed when defining the route) although the documentation does not state so. Please feel free to open a fresh issue for that, I think this is just some design issue we didn't see before 🙂

**Comment by mcollina** [comment]
> Is the rationale documented anywhere for why Fastify uses a custom request serializer instead of pino's default request serializer?

pino's default logs node core objects, while Fastify logs its own

> Regardless of support for redact use case, is the Fastify team open to logging the query string in object form?

I think this would slow things down significantly for the basic use case.

---

As a side note, I personally think including PII/sensitive information in query parameters is something that should be discouraged and a bad practice. Don't do it: those values can be leaked everywhere.

**Comment by Uzlopak** [comment]
We log the full url but you want just the path. We do this for performance reasons, as parsing the querypart takes some cycles. If we log in the onRequest stage, then we have no querystring. And why should fastify parse in that stage the query just for the logger use and then in in the parsing stage again?

Imho what you need is a way to redact the string. E.g. via a replaceAll and a regex.

**Comment by climba03003** [comment]
Though out the conversation, I believe the answer is NO here.
You should customize the logger as per your needs.
